spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java

```diff
@@ -0,0 +1,239 @@
+    this.partitionDiscoveryParallelism = spark.sessionState().conf().parallelPartitionDiscoveryParallelism();
+    this.table = table;
+    this.ops = ((HasTableOperations) table).operations();
+    this.location = table.location();
+  }
+
+  public RemoveOrphanFilesAction location(String newLocation) {
+    this.location = newLocation;
+    return this;
+  }
+
+  public RemoveOrphanFilesAction olderThan(long newOlderThanTimestamp) {
+    this.olderThanTimestamp = newOlderThanTimestamp;
+    return this;
+  }
+
+  public RemoveOrphanFilesAction deleteWith(Consumer<String> newDeleteFunc) {
+    this.deleteFunc = newDeleteFunc;
+    return this;
+  }
+
+  @Override
+  public List<String> execute() {
+    Preconditions.checkArgument(olderThanTimestamp != null, "olderThanTimestamp must be set");
+
+    Dataset<Row> validDataFileDF = buildValidDataFileDF();
+    Dataset<Row> validMetadataFileDF = buildValidMetadataFileDF();
+    Dataset<Row> validFileDF = validDataFileDF.union(validMetadataFileDF);
+    Dataset<Row> actualFileDF = buildActualFileDF();
+
+    Column joinCond = validFileDF.col("file_path").equalTo(actualFileDF.col("file_path"));
+    List<String> orphanFiles = actualFileDF.join(validFileDF, joinCond, "leftanti")
+        .as(Encoders.STRING())
+        .collectAsList();
+
+    Tasks.foreach(orphanFiles)
+        .noRetry()
+        .suppressFailureWhenFinished()
+        .onFailure((file, exc) -> LOG.warn("Failed to delete file: {}", file, exc))
+        .run(deleteFunc::accept);
+
+    return orphanFiles;
+  }
+
+  private Dataset<Row> buildValidDataFileDF() {
+    String allDataFilesMetadataTable = metadataTableName(MetadataTableType.ALL_DATA_FILES);
+    return spark.read().format("iceberg")
+        .load(allDataFilesMetadataTable)
+        .select("file_path");
+  }
+
+  private Dataset<Row> buildValidMetadataFileDF() {
+    String allManifestsMetadataTable = metadataTableName(MetadataTableType.ALL_MANIFESTS);
+    Dataset<Row> manifestDF = spark.read().format("iceberg")
+        .load(allManifestsMetadataTable)
+        .selectExpr("path as file_path");
+
+    List<String> otherMetadataFiles = Lists.newArrayList();
+
+    for (Snapshot snapshot : table.snapshots()) {
+      String manifestListLocation = snapshot.manifestListLocation();
+      if (manifestListLocation != null) {
+        otherMetadataFiles.add(manifestListLocation);
+      }
+    }
+
+    otherMetadataFiles.add(ops.metadataFileLocation("version-hint.text"));
+
+    TableMetadata metadata = ops.current();
+    otherMetadataFiles.add(metadata.file().location());
+    for (TableMetadata.MetadataLogEntry previousMetadataFile : metadata.previousFiles()) {
+      otherMetadataFiles.add(previousMetadataFile.file());
+    }
+
+    Dataset<Row> otherMetadataFileDF = spark
+        .createDataset(otherMetadataFiles, Encoders.STRING())
+        .toDF("file_path");
+
+    return manifestDF.union(otherMetadataFileDF);
+  }
+
+  private Dataset<Row> buildActualFileDF() {
```

Comment:
a comment here on criteria for building the list of actual files would be helpful.

Reply:
@prodeezy, do you mean the actual algorithm or that we select files older than a given timestamp?